pkg/check/privilege.go

```diff
@@ -0,0 +1,122 @@
+package check
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"strings"
+
+	"github.com/juju/errors"
+	"github.com/ngaut/log"
+	"github.com/pingcap/tidb-tools/pkg/dbutil"
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/parser"
+)
+
+/*****************************************************/
+
+// SourcePrivilegeChecker checks data source privileges.
+type SourcePrivilegeChecker struct {
+	db     *sql.DB
+	dbinfo *dbutil.DBConfig
+}
+
+// NewSourcePrivilegeChecker returns a Checker.
+func NewSourcePrivilegeChecker(db *sql.DB, dbinfo *dbutil.DBConfig) Checker {
+	return &SourcePrivilegeChecker{db: db, dbinfo: dbinfo}
+}
+
+// Check implements the Checker interface.
+// We only check REPLICATION SLAVE, REPLICATION CLIENT, RELOAD privileges.
+// REPLICATION SLAVE and REPLICATION CLIENT are required.
+// RELOAD is strongly suggested to have.
+func (pc *SourcePrivilegeChecker) Check(ctx context.Context) *Result {
+	result := &Result{
+		Name:  pc.Name(),
+		Desc:  "checks data source privileges",
+		State: StateFailure,
+		Extra: fmt.Sprintf("%s:%d", pc.dbinfo.Host, pc.dbinfo.Port),
+	}
+
+	grants, err := dbutil.ShowGrants(ctx, pc.db, "", "")
```

Comment:
should we specify the user?

Reply:
no need, we check current user/host (that are specified while we create db connection)